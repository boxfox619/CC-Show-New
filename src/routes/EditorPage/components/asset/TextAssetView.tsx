import * as React from 'react';
import styled from 'styled-components';

const EditorContainer = styled.div`
    height: 100%;
    overflow: hidden;
    outline: none;
`

declare let CKEDITOR: any;

interface OwnProps {
    assetId: any,
    controllable: boolean,
    value: string,
    editing: boolean,
    handleChange: (value: string) => void
}
type Props = OwnProps & React.HTMLAttributes<HTMLDivElement>;

const config = {
    toolbar: [
        { name: 'styles', items: ['Font', 'FontSize'] },
        { name: 'basicstyles', items: ['Bold', 'Italic', 'Underline', 'Strike'] },
        { name: 'paragraph', items: ['JustifyLeft', 'JustifyCenter', 'JustifyRight', 'JustifyBlock'] },
        { name: 'colors', items: ['TextColor', 'BGColor'] }
    ]
};

export default class TextAssetView extends React.Component<Props>{
    editorRef: React.RefObject<HTMLDivElement>;

    constructor(props: Props){
        super(props);
        this.editorRef = React.createRef();
    }

    public render() {
        const { assetId, controllable, editing, value, handleChange, ...divProps } = this.props;
        return (
            <EditorContainer ref={this.editorRef}
                {...divProps}
                contentEditable={controllable}
                dangerouslySetInnerHTML={{ __html: value }}
                id={this.id}
            />
        )
    }

    componentDidMount() {
        const { controllable, handleChange } = this.props;
        if (controllable && typeof CKEDITOR !== "undefined") {
            const instance = CKEDITOR.inline(this.id, config);
            const changeHandler = () => {
                const data = instance.getData();
                handleChange(data);
            };
            instance.on("change", changeHandler);
            CKEDITOR.on("instanceReady", (ck: any) => {
                ck.editor.removeMenuItem("paste");
            });
            CKEDITOR.disableAutoInline = true;
        }
    }

    shouldComponentUpdate(nextProps: Props) {
        const element = this.editorRef.current;
        if(!element){
            return false;
        }
        if (nextProps.editing) {
            element.focus();
        } else {
            element.blur();
        }
        const instance = CKEDITOR.instances[this.id];
        const data = instance.getData();
        return nextProps.value !== data || nextProps.editing !== this.props.editing;
      }

    get id() {
        return `text-asset-${this.props.assetId}`
    }
}