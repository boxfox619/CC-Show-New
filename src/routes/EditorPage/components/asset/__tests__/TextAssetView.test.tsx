import * as React from 'react';
import * as Enzyme from 'enzyme';
import TextAssetView from '../TextAssetView';

describe('<TextAssetView />', () => {
    const mockFn = jest.fn();
    it('should match snapshot', () => {
        const wrapper = Enzyme.shallow(<TextAssetView
            assetId={123}
            controllable={false}
            value="text"
            editing={false}
            handleChange={mockFn} />);
        expect(wrapper).toMatchSnapshot();
    });
});