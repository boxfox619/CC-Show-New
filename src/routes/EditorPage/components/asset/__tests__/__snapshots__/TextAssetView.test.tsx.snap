// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextAssetView /> should match snapshot 1`] = `ShallowWrapper {}`;
